### PR TITLE
[`Research Project`] Add AnyText: Multilingual Visual Text Generation And Editing

### LINKED ISSUE #6407: AnyText: Multilingual Visual Text Generation And Editing
### Model/Pipeline/Scheduler description

From the [repository](https://github.com/tyxsspa/AnyText): 

> AnyText comprises a diffusion pipeline with two primary elements: an auxiliary latent module and a text embedding module. The former uses inputs like text glyph, position, and masked image to generate latent features for text generation or editing. The latter employs an OCR model for encoding stroke data as embeddings, which blend with image caption embeddings from the tokenizer to generate texts that seamlessly integrate with the background. We employed text-control diffusion loss and text perceptual loss for training to further enhance writing accuracy.

image

### Open source status

- [X] The model implementation is available.
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Repository: https://github.com/tyxsspa/AnyText

Paper: https://arxiv.org/abs/2311.03054

Weights and inference code: https://modelscope.cn/models/damo/cv_anytext_text_generation_editing/summary

### PR DESCRIPTION
**Pull request by tolgacangoz**
Thanks for the opportunity to fix #6407!

>AnyText comprises a diffusion pipeline with two primary elements: an auxiliary latent module and a text embedding module. The former uses inputs like text glyph, position, and masked image to generate latent features for text generation or editing. The latter employs an OCR model for encoding stroke data as embeddings, which blend with image caption embeddings from the tokenizer to generate texts that seamlessly integrate with the background. We employed text-control diffusion loss and text perceptual loss for training to further enhance writing accuracy.

Project page: https://aigcdesigngroup.github.io/homepage_anytext

anytext
anytext

TODOs:
:white_check_mark: `AuxiliaryLatentModule`
:white_check_mark: `AnyTextControlNetModel` -> Inherited and adapted from `ControlNetModel`. The only difference is that using Gylph Block, Position Block, and Fuse Block instead of `input_hint_block` or `controlnet_cond_embedding` from an ordinary ControlNet -`ControlNetConditioningEmbedding` is different.
:white_check_mark: `AnyTextPipeline` -> Adapted from `StableDiffusionControlNetPipeline`.
:white_check_mark: `TextEmbeddingModule` -> Replaces the `encode_prompt()` function. I may transfer what `TextEmbeddingModule` does into `encode_prompt()`.
:grey_question: `convert_anytext_to_diffusers.py` -> I didn't collect them in a single file/command, so I didn't include it here. Is it really necessary for now?
:white_check_mark: Finish HF integration & upload converted checkpoints to HF.
:white_check_mark: `README.md`

[Open In Colab](https://colab.research.google.com/gist/tolgacangoz/b87ec9d2f265b448dd947c9d4a0da389/anytext.ipynb)

I want to congratulate you for this great work @tyxsspa, [Wangmeng Xiang](https://openreview.net/profile?id=~Wangmeng_Xiang1), [Jun-Yan He](https://openreview.net/profile?id=~Jun-Yan_He2), @gengyifeng, @xungie, and thank you for open-sourcing with the Apache 2.0 License!

<details>
<summary>To be continued...?</summary>

4-Figure2-1

</details>

@sayakpaul @yiyixuxu @a-r-r-o-w @asomoza

### CODE REVIEW

**Comment by tolgacangoz** [comment]
<details><summary>And the first result is ...
</summary>

 Original Implementation            |  My Current Implementation
:-------------------------:|:-------------------------:
| original | anytext_mine |

... really bad 😅. There is obviously something wrong. Now, I am working on this.

</details>

**Comment by a-r-r-o-w** [comment]
Absolutely amazing work here @tolgacangoz :heart: Thank you for picking this up after a stream of contributors (including me) mentioned that they'd take it up but weren't able to! The PR looks mostly good to me but I'll wait for it to be unmarked from draft. 

Really really cool work here! cc @sayakpaul

**Comment by tolgacangoz** [comment]
Thanks @a-r-r-o-w!

This PR is almost done. I am currently working on the Matryoshka model. I guess it is the priority. When I complete it I will turn back to this PR.

**Comment by tolgacangoz** [comment]
Thanks so much @a-r-r-o-w!

This PR is almost done. I am currently working on the Matryoshka model. I guess it is the priority. When I complete it I will turn back to this PR.

**Comment by tolgacangoz** [comment]
Because I have had a full time job for 1 month, I couldn't return to this PR. I will try to complete this. Last giant models seem very good for text, but this model still might be preferable because of its size etc.

**Comment by tolgacangoz** [comment]
Sorry for the delay, I was busy with other priorities. This PR will be completed within 1 week.

**Comment by tolgacangoz** [comment]
not stale

**Comment by asomoza** [comment]
Hi @tolgacangoz, are you still working on this?

**Comment by tolgacangoz** [comment]
Hi Álvaro. My priorities have had to change over the last 4-5 months. Starting tomorrow, I plan to complete this PR within 1 week.

**Comment by tolgacangoz** [comment]
So, what now?

**Comment by tolgacangoz** [comment]
This will be my second pipeline contribution, yay :partying_face:

**Comment by asomoza** [comment]
thanks a lot, it looks good to me, really amazing project and port to diffusers with good results.

anytext

ccing @a-r-r-o-w because of https://github.com/huggingface/diffusers/pull/8998#issuecomment-2308036940

File: examples/research_projects/anytext/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+
+```py
+import torch
+from diffusers import DiffusionPipeline
+from anytext_controlnet import AnyTextControlNetModel
+from diffusers.utils import load_image
+
+# I chose a font file shared by an HF staff:
+!wget https://huggingface.co/spaces/ysharma/TranslateQuotesInImageForwards/resolve/main/arial-unicode-ms.ttf
```
Comment (a-r-r-o-w) [inline examples/research_projects/anytext/README.md:18]:
```suggestion
# !wget https://huggingface.co/spaces/ysharma/TranslateQuotesInImageForwards/resolve/main/arial-unicode-ms.ttf
```

Could we format these example in ruff format btw for consistency

File: examples/research_projects/anytext/anytext.py [hunk 2]
```diff
@@ -0,0 +1,2397 @@
...
+                attention_mask=attention_mask,
+                causal_attention_mask=causal_attention_mask,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+            )
+            last_hidden_state = self.final_layer_norm(last_hidden_state)
+            return last_hidden_state
+
+        self.transformer.text_model.forward = text_encoder_forward.__get__(self.transformer.text_model)
```
Comment (a-r-r-o-w) [inline examples/research_projects/anytext/anytext.py:652]: This is not really in style of our format, but since it's a research-project addition, we should be good (just making a note for other reviewers and not really a block IMO)

File: examples/research_projects/anytext/anytext.py [hunk 3]
```diff
@@ -0,0 +1,2397 @@
...
+        if remaining_columns > 0:
+            remaining_group = trimmed_encoding[:, -remaining_columns:]
+            padding_columns = chunk_size - remaining_group.shape[1]
+            padding = id_end.expand(bs, padding_columns)
+            remaining_group_pad = torch.cat((id_start, remaining_group, padding, id_end), dim=1)
+            tokens_list.append(remaining_group_pad)
+        return tokens_list
+
+
+class TextEmbeddingModule(ModelMixin, ConfigMixin):
```
Comment (a-r-r-o-w) [inline examples/research_projects/anytext/anytext.py:729]: The reason for adding `ModelMixin, ConfigMixin` is that modules that have it should directly supporting loading its state dict using `.from_pretrained`. Many classes seem to have that here but as it is a research-project addition, it is not necessary to follow the general design philosophies IMO (to reduce overhead to amazing contributors like yourself!)

File: examples/research_projects/anytext/anytext.py [hunk 4]
```diff
@@ -0,0 +1,2397 @@
...
+            timestep_cond = self.get_guidance_scale_embedding(
+                guidance_scale_tensor, embedding_dim=self.unet.config.time_cond_proj_dim
+            ).to(device=device, dtype=latents.dtype)
+
+        # 4. Prepare image
+        if isinstance(controlnet, ControlNetModel):
+            # image = self.prepare_image(
+            #     image=image,
+            #     width=width,
+            #     height=height,
```
Comment (a-r-r-o-w) [inline examples/research_projects/anytext/anytext.py:2190]: Should the commented lines be removed in some places?

**Review by a-r-r-o-w — APPROVED** [review]
Thanks @tolgacangoz! I have been checking the progress on this PR on and off for a long while :) Really amazing work and it's something very rare to see from contributor! Sincerely thanks a lot for your super hard work adding this sota text generation method (atleast it was when it came out)

As it's a research-project addition, and works well based on Alvaro's comment, I believe it's completely fine to not strictly follow the diffusers-format philosophies about modeling/pipeline exactly. @yiyixuxu can take a final look and merge 🤗

File: examples/research_projects/anytext/README.md (hunk 1 above)
Comment (tolgacangoz) [reply examples/research_projects/anytext/README.md]: Do you mean I am supposed to run `make style && make quality` with the latest version of ruff?

File: examples/research_projects/anytext/anytext.py (hunk 2 above)
Comment (tolgacangoz) [reply examples/research_projects/anytext/anytext.py:652]: Yes, I didn't consider such things.

File: examples/research_projects/anytext/README.md (hunk 1 above)
Comment (a-r-r-o-w) [reply examples/research_projects/anytext/README.md]: Yeah, so we don't have a way, at the moment, to format the code examples in README's, so they can't be formatted. It's alright though, and not a blocker for merging.

Thanks for your patience and working on this amazing integration!

**Review by a-r-r-o-w — APPROVED** [review]
Thanks @tolgacangoz!

**Comment by tolgacangoz** [comment]
Thanks for this opportunity to contribute!
